project-koku/koku
# Handling 404 when setting source status

Ensure that we do not throw an exception when attempting to set a source status on an application object that has been removed.

**Testing**
1. Create a GCP source with invalid project so that the source is in an unavailable state.
2. Delete the application object in the sources-api
3. Verify that no exception is thrown while set_source_status is called

**Test Results**
```
sources_client    | [2021-03-29 17:19:27,444] INFO None Provider account validation started for {'dataset': 'billing_datset'}.
sources_client    | [2021-03-29 17:19:27,980] WARNING None Improper IAM permissions: [].
sources_client    | [2021-03-29 17:19:27,985] WARNING None Unable to create provider for Source ID: 1. Reason: {'gcp.iam.permissions': [ErrorDetail(string='Incorrect IAM permissions for project wrong', code='invalid')]}
sources_client    | [2021-03-29 17:19:28,020] INFO None validation error: {'gcp.iam.permissions': [ErrorDetail(string='Incorrect IAM permissions for project wrong', code='invalid')]}. Validation detail {'gcp.iam.permissions': [ErrorDetail(string='Incorrect IAM permissions for project wrong', code='invalid')]}
sources_client    | [2021-03-29 17:19:28,020] WARNING None Source ID: 1 error message: Incorrect IAM permissions for project wrong
sources_client    | [2021-03-29 17:19:28,024] INFO None Setting Source Status for Source ID: 1: {'availability_status': 'unavailable', 'availability_status_error': 'Incorrect IAM permissions for project wrong'}
sources_client    | [2021-03-29 17:19:28,150] INFO None Koku provider operation to execute: create for Source ID: 1 complete.
sources_client    | [2021-03-29 17:19:28,155] INFO None Authentication attached to Source ID: 1
sources_client    | [2021-03-29 17:19:28,184] INFO None Processing message offset: 52 partition: 0
sources_client    | [2021-03-29 17:19:28,184] INFO None Cost Management Message to process: {'event_type': 'Application.update', 'offset': 52, 'partition': 0, 'source_id': 1, 'auth_header': 'eyJpZGVudGl0eSI6IHsiYWNjb3VudF9udW1iZXIiOiAiMTAwMDEiLCAidHlwZSI6ICJVc2VyIiwgInVzZXIiOiB7InVzZXJuYW1lIjogImNvc3QtbWdtdCIsICJlbWFpbCI6ICJjb3N0LW1nbXRAcmVkaGF0LmNvbSIsICJpc19vcmdfYWRtaW4iOiB0cnVlfX19'}
sources_client    | [2021-03-29 17:19:28,301] INFO None Authentication attached to Source ID: 1
sources_client    | [2021-03-29 17:19:28,301] INFO None Found settings: {'billing_source': {'data_source': {'dataset': 'billing_datset'}}}
sources_client    | [2021-03-29 17:20:21,863] INFO None Processing message offset: 53 partition: 0
sources_client    | [2021-03-29 17:20:21,863] INFO None Cost Management Message to process: {'event_type': 'Application.destroy', 'offset': 53, 'partition': 0, 'source_id': 1, 'auth_header': 'eyJpZGVudGl0eSI6IHsiYWNjb3VudF9udW1iZXIiOiAiMTAwMDEiLCAidHlwZSI6ICJVc2VyIiwgInVzZXIiOiB7InVzZXJuYW1lIjogInVzZXJfZGV2IiwgImVtYWlsIjogInVzZXJfZGV2QGZvby5jb20iLCAiaXNfb3JnX2FkbWluIjogdHJ1ZX19LCAiZW50aXRsZW1lbnRzIjogeyJjb3N0X21hbmFnZW1lbnQiOiB7ImlzX2VudGl0bGVkIjogdHJ1ZX19fQ=='}
sources_client    | [2021-03-29 17:20:21,869] INFO None Adding operation destroy for GCP Source to process queue (size: 0)
sources_client    | [2021-03-29 17:20:21,870] INFO None Koku provider operation to execute: destroy for Source ID: 1
sources_client    | [2021-03-29 17:20:21,923] INFO None Destroy provider task queued for provider None for Source ID: 1
sources_client    | [2021-03-29 17:20:22,024] INFO None Koku provider operation to execute: destroy for Source ID: 1 complete.
sources_client    | [2021-03-29 17:20:22,029] INFO None Provider None does not exist.  Unable to mark as inactive

koku-worker       | [2021-03-29 17:20:21,934] INFO b262492e-b685-4d09-9057-1dfa213174ea Deleting Provider None for Source ID: 1
koku-worker       | [2021-03-29 17:20:21,934: INFO/ForkPoolWorker-1] Deleting Provider None for Source ID: 1
koku-worker       | [2021-03-29 17:20:21,982] INFO b262492e-b685-4d09-9057-1dfa213174ea Provider does not exist, skipping Provider delete.
koku-worker       | [2021-03-29 17:20:21,982: INFO/ForkPoolWorker-1] Provider does not exist, skipping Provider delete.
koku-worker       | [2021-03-29 17:20:22,040] INFO b262492e-b685-4d09-9057-1dfa213174ea source.storage.destroy_source_event destroyed Source ID: 1
koku-worker       | [2021-03-29 17:20:22,040: INFO/ForkPoolWorker-1] source.storage.destroy_source_event destroyed Source ID: 1
```
